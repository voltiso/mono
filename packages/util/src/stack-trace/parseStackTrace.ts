// â €â“¥ 2024     ðŸŒ©    ðŸŒ©     â €   â €
// â €         ðŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €

import type { UndefinedFromOptional } from '~/object'
import { define } from '~/type'

import { tryParseCodeLocation } from './parseCodeLocation'

export interface StackTraceEntry
	extends UndefinedFromOptional<{
		functionName: string

		path?: string
		fileName?: string
		line?: number
		column?: number
	}> {}

export function parseStackTrace(stackStr: string): StackTraceEntry[] {
	const entries = stackStr
		.split('\n')
		.map(
			line =>
				/at (?<functionName>(?:new )?[^ ]*) [^(]*\((?<location>[^)]*)\)/u.exec(
					line,
				)?.groups as
					| {
							functionName: string
							location: string
					  }
					| undefined,
		)
		.filter((entry): entry is typeof entry & object => !!entry)
		.map(({ functionName, location }) => {
			const parsed = tryParseCodeLocation(location)

			return define<StackTraceEntry>().value({
				functionName,
				...parsed,
			})
		})

	return entries
}
