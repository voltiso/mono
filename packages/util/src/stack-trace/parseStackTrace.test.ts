// â €â“¥ 2024     ðŸŒ©    ðŸŒ©     â €   â €
// â €         ðŸŒ© VÍ›oÍ›Í›Í›ltÍ›Í›Í›iÍ›Í›Í›Í›soÍ›Í›Í›.comâ €  â €â €â €

import { parseStackTrace } from './parseStackTrace'

describe('parseStackTrace', () => {
	it('shell stack', () => {
		expect.hasAssertions()

		const stackStr =
			'Error\n' +
			'    at REPL6:1:1\n' +
			'    at Script.runInThisContext (node:vm:129:12)\n' +
			'    at REPLServer.defaultEval (node:repl:572:29)\n' +
			'    at bound (node:domain:433:15)\n' +
			'    at REPLServer.runBound [as eval] (node:domain:444:12)\n' +
			'    at REPLServer.onLine (node:repl:902:10)\n' +
			'    at REPLServer.emit (node:events:525:35)\n' +
			'    at REPLServer.emit (node:domain:489:12)\n' +
			'    at REPLServer.Interface._onLine (node:readline:491:10)\n' +
			'    at REPLServer.Interface._line (node:readline:868:8)'

		const stack = parseStackTrace(stackStr)

		expect(stack).toMatchSnapshot()
	})

	it('jest stack', () => {
		expect.hasAssertions()

		const stackStr = `Error: 
        at Object.<anonymous> (/home/atablash/mono/packages/util/src/stack-trace/parseStackTrace.test.ts:22:25)
        at Promise.then.completed (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/utils.js:289:28)
        at new Promise (<anonymous>)
        at callAsyncCircusFn (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/utils.js:222:10)
        at _callCircusTest (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js:248:40)
        at _runTest (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js:184:3)
        at _runTestsForDescribeBlock (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js:86:9)
        at _runTestsForDescribeBlock (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js:81:9)
        at run (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js:26:3)
        at runAndTransformResultsToJestFormat (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapterInit.js:120:21)
        at jestAdapter (/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:79:19)
        at runTestInternal (/home/atablash/mono/node_modules/.pnpm/jest-runner@29.2.0/node_modules/jest-runner/build/runTest.js:367:16)
        at runTest (/home/atablash/mono/node_modules/.pnpm/jest-runner@29.2.0/node_modules/jest-runner/build/runTest.js:444:34)`

		const stack = parseStackTrace(stackStr)

		expect(stack).toMatchSnapshot()
	})
})
