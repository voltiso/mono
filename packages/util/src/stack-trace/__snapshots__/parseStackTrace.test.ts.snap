// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseStackTrace jest stack 1`] = `
[
  {
    "column": 25,
    "functionName": "Object.<anonymous>",
    "line": 22,
    "path": "/home/atablash/mono/packages/util/src/stack-trace/parseStackTrace.test.ts",
  },
  {
    "column": 28,
    "functionName": "Promise.then.completed",
    "line": 289,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/utils.js",
  },
  {
    "functionName": "new Promise",
    "path": "<anonymous>",
  },
  {
    "column": 10,
    "functionName": "callAsyncCircusFn",
    "line": 222,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/utils.js",
  },
  {
    "column": 40,
    "functionName": "_callCircusTest",
    "line": 248,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js",
  },
  {
    "column": 3,
    "functionName": "_runTest",
    "line": 184,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js",
  },
  {
    "column": 9,
    "functionName": "_runTestsForDescribeBlock",
    "line": 86,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js",
  },
  {
    "column": 9,
    "functionName": "_runTestsForDescribeBlock",
    "line": 81,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js",
  },
  {
    "column": 3,
    "functionName": "run",
    "line": 26,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/run.js",
  },
  {
    "column": 21,
    "functionName": "runAndTransformResultsToJestFormat",
    "line": 120,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapterInit.js",
  },
  {
    "column": 19,
    "functionName": "jestAdapter",
    "line": 79,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-circus@29.2.0/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js",
  },
  {
    "column": 16,
    "functionName": "runTestInternal",
    "line": 367,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-runner@29.2.0/node_modules/jest-runner/build/runTest.js",
  },
  {
    "column": 34,
    "functionName": "runTest",
    "line": 444,
    "path": "/home/atablash/mono/node_modules/.pnpm/jest-runner@29.2.0/node_modules/jest-runner/build/runTest.js",
  },
]
`;

exports[`parseStackTrace shell stack 1`] = `
[
  {
    "column": 12,
    "functionName": "Script.runInThisContext",
    "line": 129,
    "path": "node:vm",
  },
  {
    "column": 29,
    "functionName": "REPLServer.defaultEval",
    "line": 572,
    "path": "node:repl",
  },
  {
    "column": 15,
    "functionName": "bound",
    "line": 433,
    "path": "node:domain",
  },
  {
    "column": 12,
    "functionName": "REPLServer.runBound",
    "line": 444,
    "path": "node:domain",
  },
  {
    "column": 10,
    "functionName": "REPLServer.onLine",
    "line": 902,
    "path": "node:repl",
  },
  {
    "column": 35,
    "functionName": "REPLServer.emit",
    "line": 525,
    "path": "node:events",
  },
  {
    "column": 12,
    "functionName": "REPLServer.emit",
    "line": 489,
    "path": "node:domain",
  },
  {
    "column": 10,
    "functionName": "REPLServer.Interface._onLine",
    "line": 491,
    "path": "node:readline",
  },
  {
    "column": 8,
    "functionName": "REPLServer.Interface._line",
    "line": 868,
    "path": "node:readline",
  },
]
`;
